alf/algorithms/misc_algorithm.py

```diff
@@ -0,0 +1,263 @@
+
+import gin.tf
+import tensorflow as tf
+
+from tf_agents.networks.network import Network
+import tf_agents.specs.tensor_spec as tensor_spec
+
+from alf.algorithms.algorithm import Algorithm, AlgorithmStep, LossInfo
+from alf.utils.data_buffer import DataBuffer
+from tf_agents.trajectories.time_step import StepType
+from alf.utils.conditional_ops import conditional_update, run_if
+
+MISCInfo = namedtuple("MISCInfo", ["reward", "loss"])
+
+
+@gin.configurable
+class MISCAlgorithm(Algorithm):
+    """Mutual Information-based State Control (MISC)
+    Author: Rui Zhao 
+    Work done during a research internship at Horizon Robotics. 
+    The paper is currently under review in a conference.
+
+    This algorithm generates the intrinsic reward based on the mutual information estimation between the states of interests and the context states.
+
+    See Zhao et al "Self-Supervised State-Control through Intrinsic Mutual Information Rewards"
+    """
+
+    def __init__(self,
+                 feature_spec,
+                 traj_spec,
+                 soi_spec,
+                 soc_spec,
+                 mi_r_scale=5000,
+                 hidden_size=256,
+                 misc_layerx: Network = None,
+                 misc_layery: Network = None,
+                 misc_layero: Network = None,
+                 buffer_size=100,
+                 n_objects=1,
+                 empowerment=False,
+                 env_name='SocialBot-PlayGround-v0',
+                 name="MISCAlgorithm"):
+        """Create an MISCAlgorithm.
+        
+        Args:
+            feature_spec: size of the state and the action
+            traj_spec: (num_parallel_environments, feature_spec)
+            soi_spec: (None, None, state of interest size)
+            soc_spec: (None, None, state of context size)
+            mi_r_scale: scale factor of mi estimation
+            hidden_size: number of hidden units in neural nets
+            misc_layerx: input layer with context states as input, output dimension is of hidden_size
+            misc_layery: input layer with states of interest as input, output dimension is of hidden_size
+            misc_layero: output layer with with input shape (None, None, hidden_size), output dimension is of 1
+            buffer_size: buffer size for the data buffer storing the trajectories for training the Mutual Information Neural Estimator
+            n_objects: number of objects for estimating the mutual information
+            empowerment: if empowerment is true, then use action instead of the context states
+            name: the algorithm name, "MISCAlgorithm"
+        """
+
+        super(MISCAlgorithm, self).__init__(
+            train_state_spec=feature_spec, name=name)
+
+        feature_dim = tf.nest.flatten(feature_spec)[0].shape[-1]
+
+        if misc_layerx is None:
+            misc_layerx = tf.keras.layers.Dense(
+                hidden_size, input_shape=soc_spec.shape, activation=None)
```

Comment:
you don't need to provide `input_shape` for `Dense` layer.

Reply:
ok